File: apps/web/src/model/geom/arc.ts
import { op } from '../../wasm/core';
import type { Vec2 } from '../geometry';

/** Circles and circular arcs, computed by the geometry core (docs/adr/0008). */

/** 2π (the double `Math.PI * 2` gives). */
export const TAU = 6.283185307179586;

/** Normalizes an angle to [0, 2π). */
export const normAngle = op<(a: number) => number>('normAngle');

/** CCW sweep from a0 to a1 in (0, 2π]. Equal angles mean a full turn. */
export const sweep = op<(a0: number, a1: number) => number>('sweep');

/** Whether angle θ lies on the CCW arc from a0 spanning `sw` (inclusive, with tolerance; 1e-9 by default). */
export const onArc = op<(theta: number, a0: number, sw: number, eps?: number) => boolean>('onArc');

/** Position of angle θ along the arc as 0..1 (for trimming parameters). */
export const arcParam = op<(theta: number, a0: number, sw: number) => number>('arcParam');

export const pointOnCircle = op<(c: Vec2, r: number, a: number) => Vec2>('pointOnCircle');

export const circleThrough = op<(p1: Vec2, p2: Vec2, p3: Vec2) => { c: Vec2; r: number } | null>('circleThrough');

export interface ArcGeom {
  c: Vec2;
  r: number;
  /** Start angle (radians, CCW from east). */
  a0: number;
  /** End angle; the arc runs CCW from a0 to a1. */
  a1: number;
}

/**
 * Arc from start through mid to end. Stored CCW, so a clockwise pick order
 * swaps start and end — the drawn shape is the same.
 */
export const arcThrough = op<(start: Vec2, mid: Vec2, end: Vec2) => ArcGeom | null>('arcThrough');

/** Points along the arc, steps of at most `maxStep` radians (a 72nd of a turn by default). */
export const tessellateArc = op<(a: ArcGeom, maxStep?: number) => Vec2[]>('tessellateArc');

export const arcStart = op<(a: ArcGeom) => Vec2>('arcStart');
export const arcEnd = op<(a: ArcGeom) => Vec2>('arcEnd');
export const arcMid = op<(a: ArcGeom) => Vec2>('arcMid');
export const arcLength = op<(a: ArcGeom) => number>('arcLength');

/** DXF bulge (tan(θ/4)) of a counter-clockwise arc. */
export const bulgeFromArc = op<(a: ArcGeom) => number>('bulgeFromArc');

File: crates/shared/geometry-core/src/geom/arc.rs
//! Angles, circles and arcs (`apps/web/src/model/geom/arc.ts`).

use crate::api::Op;
use crate::jsmath::{TAU, atan2, cos, js_hypot, js_max, sin, tan};
use crate::op;
use crate::vec2::Vec2;

/// Normalizes an angle to [0, 2π).
pub fn norm_angle(a: f64) -> f64 {
    let r = a % TAU;
    if r < 0.0 { r + TAU } else { r }
}

/// CCW sweep from a0 to a1 in (0, 2π]; equal angles mean a full turn.
pub fn sweep(a0: f64, a1: f64) -> f64 {
    let s = norm_angle(a1 - a0);
    if s < 1e-12 { TAU } else { s }
}

/// Whether θ lies on the CCW arc from a0 spanning `sw` (inclusive, with tolerance).
pub fn on_arc(theta: f64, a0: f64, sw: f64, eps: f64) -> bool {
    if sw >= TAU - eps {
        return true;
    }
    let d = norm_angle(theta - a0);
    d <= sw + eps || d >= TAU - eps
}

pub const ON_ARC_EPS: f64 = 1e-9;

/// Position of θ along the arc as 0..1.
pub fn arc_param(theta: f64, a0: f64, sw: f64) -> f64 {
    norm_angle(theta - a0) / sw
}

pub fn point_on_circle(c: Vec2, r: f64, a: f64) -> Vec2 {
    Vec2::new(c.x + cos(a) * r, c.y + sin(a) * r)
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Circle {
    pub c: Vec2,
    pub r: f64,
}

crate::json_struct!(Circle { c, r });

pub fn circle_through(p1: Vec2, p2: Vec2, p3: Vec2) -> Option<Circle> {
    let ax = p2.x - p1.x;
    let ay = p2.y - p1.y;
    let bx = p3.x - p1.x;
    let by = p3.y - p1.y;
    let d = 2.0 * (ax * by - ay * bx);
    if d.abs() < 1e-12 {
        return None; // collinear
    }
    let a2 = ax * ax + ay * ay;
    let b2 = bx * bx + by * by;
    let cx = (by * a2 - ay * b2) / d;
    let cy = (ax * b2 - bx * a2) / d;
    Some(Circle {
        c: Vec2::new(p1.x + cx, p1.y + cy),
        r: js_hypot(cx, cy),
    })
}

/// An arc running counter-clockwise from `a0` to `a1` (radians from east).
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct ArcGeom {
    pub c: Vec2,
    pub r: f64,
    pub a0: f64,
    pub a1: f64,
}

crate::json_struct!(ArcGeom { c, r, a0, a1 });

/// Arc from start through mid to end, stored CCW (a clockwise pick order swaps the ends).
pub fn arc_through(start: Vec2, mid: Vec2, end: Vec2) -> Option<ArcGeom> {
    let Circle { c, r } = circle_through(start, mid, end)?;
    let a_s = atan2(start.y - c.y, start.x - c.x);
    let a_m = atan2(mid.y - c.y, mid.x - c.x);
    let a_e = atan2(end.y - c.y, end.x - c.x);
    let ccw = norm_angle(a_m - a_s) < norm_angle(a_e - a_s);
    Some(if ccw {
        ArcGeom {
            c,
            r,
            a0: norm_angle(a_s),
            a1: norm_angle(a_e),
        }
    } else {
        ArcGeom {
            c,
            r,
            a0: norm_angle(a_e),
            a1: norm_angle(a_s),
        }
    })
}

pub const DEFAULT_STEP: f64 = TAU / 72.0;

pub fn tessellate_arc(a: &ArcGeom, max_step: f64) -> Vec<Vec2> {
    let sw = sweep(a.a0, a.a1);
    let n = js_max(2.0, (sw / max_step).ceil());
    let mut pts = Vec::new();
    let mut i = 0.0;
    while i <= n {
        pts.push(point_on_circle(a.c, a.r, a.a0 + (sw * i) / n));
        i += 1.0;
    }
    pts
}

pub fn arc_start(a: &ArcGeom) -> Vec2 {
    point_on_circle(a.c, a.r, a.a0)
}
pub fn arc_end(a: &ArcGeom) -> Vec2 {
    point_on_circle(a.c, a.r, a.a1)
}
pub fn arc_mid(a: &ArcGeom) -> Vec2 {
    point_on_circle(a.c, a.r, a.a0 + sweep(a.a0, a.a1) / 2.0)
}
pub fn arc_length(a: &ArcGeom) -> f64 {
    a.r * sweep(a.a0, a.a1)
}
/// DXF bulge (tan(θ/4)) of a counter-clockwise arc.
pub fn bulge_from_arc(a: &ArcGeom) -> f64 {
    tan(sweep(a.a0, a.a1) / 4.0)
}

pub(crate) static OPS: &[Op] = &[
    op!("normAngle", |a: f64| norm_angle(a)),
    op!("sweep", |a0: f64, a1: f64| sweep(a0, a1)),
    op!("onArc", |theta: f64, a0: f64, sw: f64, eps: Option<f64>| {
        on_arc(theta, a0, sw, eps.unwrap_or(ON_ARC_EPS))
    }),
    op!("arcParam", |theta: f64, a0: f64, sw: f64| arc_param(
        theta, a0, sw
    )),
    op!("pointOnCircle", |c: Vec2, r: f64, a: f64| {
        point_on_circle(c, r, a)
    }),
    op!("circleThrough", |p1: Vec2, p2: Vec2, p3: Vec2| {
        circle_through(p1, p2, p3)
    }),
    op!("arcThrough", |s: Vec2, m: Vec2, e: Vec2| arc_through(
        s, m, e
    )),
    op!("tessellateArc", |a: ArcGeom, step: Option<f64>| {
        tessellate_arc(&a, step.unwrap_or(DEFAULT_STEP))
    }),
    op!("arcStart", |a: ArcGeom| arc_start(&a)),
    op!("arcEnd", |a: ArcGeom| arc_end(&a)),
    op!("arcMid", |a: ArcGeom| arc_mid(&a)),
    op!("arcLength", |a: ArcGeom| arc_length(&a)),
    op!("bulgeFromArc", |a: ArcGeom| bulge_from_arc(&a)),
];
